File: experiments/gilded-rose/csharp/attempts/opus46high/3/GildedRose.cs
using System;
using System.Collections.Generic;

namespace GildedRose;

public class GildedRoseShop
{
    private readonly IList<Item> _items;

    public GildedRoseShop(IList<Item> items)
    {
        _items = items;
    }

    public void UpdateQuality()
    {
        foreach (var item in _items)
            GetStrategy(item).Update(item);
    }

    private static IItemStrategy GetStrategy(Item item) => item.Name switch
    {
        "Sulfuras, Hand of Ragnaros" => new SulfurasStrategy(),
        "Aged Brie" => new AgedBrieStrategy(),
        "Backstage passes to a TAFKAL80ETC concert" => new BackstagePassStrategy(),
        var name when name.StartsWith("Conjured", StringComparison.Ordinal) => new ConjuredStrategy(),
        _ => new NormalStrategy()
    };
}

public interface IItemStrategy
{
    void Update(Item item);
}

public class SulfurasStrategy : IItemStrategy
{
    public void Update(Item item) { }
}

public class AgedBrieStrategy : IItemStrategy
{
    public void Update(Item item)
    {
        item.SellIn--;
        item.Quality = Math.Min(50, item.Quality + (item.SellIn < 0 ? 2 : 1));
    }
}

public class BackstagePassStrategy : IItemStrategy
{
    public void Update(Item item)
    {
        item.SellIn--;
        if (item.SellIn < 0)
        {
            item.Quality = 0;
            return;
        }
        var increase = item.SellIn < 5 ? 3 : item.SellIn < 10 ? 2 : 1;
        item.Quality = Math.Min(50, item.Quality + increase);
    }
}

public class NormalStrategy : IItemStrategy
{
    public void Update(Item item)
    {
        item.SellIn--;
        var degradation = item.SellIn < 0 ? 2 : 1;
        item.Quality = Math.Max(0, item.Quality - degradation);
    }
}

public class ConjuredStrategy : IItemStrategy
{
    public void Update(Item item)
    {
        item.SellIn--;
        var degradation = item.SellIn < 0 ? 4 : 2;
        item.Quality = Math.Max(0, item.Quality - degradation);
    }
}

File: experiments/gilded-rose/fsharp/attempts/opus46high/3/GildedRose.fs
module GildedRose

open System.Collections.Generic

type Item = { Name: string; SellIn: int; Quality: int }

type GildedRoseShop(items: IList<Item>) =
    member _.Items = items

    member _.UpdateQuality() =
        for i = 0 to items.Count - 1 do
            if items.[i].Name <> "Aged Brie" && items.[i].Name <> "Backstage passes to a TAFKAL80ETC concert" then
                if items.[i].Quality > 0 then
                    if items.[i].Name <> "Sulfuras, Hand of Ragnaros" then
                        items.[i] <- { items.[i] with Quality = items.[i].Quality - 1 }
            else
                if items.[i].Quality < 50 then
                    items.[i] <- { items.[i] with Quality = items.[i].Quality + 1 }
                    if items.[i].Name = "Backstage passes to a TAFKAL80ETC concert" then
                        if items.[i].SellIn < 11 then
                            if items.[i].Quality < 50 then
                                items.[i] <- { items.[i] with Quality = items.[i].Quality + 1 }
                        if items.[i].SellIn < 6 then
                            if items.[i].Quality < 50 then
                                items.[i] <- { items.[i] with Quality = items.[i].Quality + 1 }
            if items.[i].Name <> "Sulfuras, Hand of Ragnaros" then
                items.[i] <- { items.[i] with SellIn = items.[i].SellIn - 1 }
            if items.[i].SellIn < 0 then
                if items.[i].Name <> "Aged Brie" then
                    if items.[i].Name <> "Backstage passes to a TAFKAL80ETC concert" then
                        if items.[i].Quality > 0 then
                            if items.[i].Name <> "Sulfuras, Hand of Ragnaros" then
                                items.[i] <- { items.[i] with Quality = items.[i].Quality - 1 }
                    else
                        items.[i] <- { items.[i] with Quality = items.[i].Quality - items.[i].Quality }
                else
                    if items.[i].Quality < 50 then
                        items.[i] <- { items.[i] with Quality = items.[i].Quality + 1 }
        ()
